### PR TITLE
Refactor to use embedded_io and embedded_io_async

### PR DESCRIPTION
**Pull request by m5p3nc3r**
Hi guys

This is a bit of a major refactor to use embedded_io and embedded_io_async as a consistent interface for the sync and async interfaces.  It build on the work I did in #17 and should replace that PR if this refactor is accepted.

There are some benefits of this new approach
- The Editor logic between sync and async is now identical
- There is no need for Tokio in the core noline implementation, only in the example
- The sync and no_sync (now renamed sync_editor and async_editor) are simplified, with IO logic moved into [a]sync_io

I have only worked to get the examples for std/std-sync and std/std-async-tokio working, but wanted to check in to see if this approach was going in the right direction.

There is still a lot to do:
- Update the comments to reflect the new code structure
- Update the no_std examples to use the new code structure
- Re-enable the tests

Please let me know what you think, and I will complete the work if you think this worthwhile.

### CODE REVIEW

**Comment by eivindbergem** [comment]
This is really good work! I had a quick look, and this is very much the going in the right direction, so please continue :smile:

**Comment by m5p3nc3r** [comment]
Hey guys

I am really struggling with implementing the Sync example.  The reason is that USB devices can return ```UsbError::WouldBlock```, in which case I would need to call the poll function again.  In the original example, this is worked around by lending the ```UsbDevice``` instance to the ```SerialWrapper``` class.  This feels wrong, as in a multi-function device, you really want to be calling poll in an outer loop as this really has nothing to do with the ```SerialWrapper``` functionality.

The way I am calling poll looks like this at the moment

````rust
let mut editor = loop {
        if !usb_dev.poll(&mut [io.inner().serial])
            || !io.inner().serial.dtr()
            || !io.inner().serial.rts()
        {
            continue;
        }

        break EditorBuilder::new_static::<128>()
            .with_static_history::<128>()
            .build_sync(&mut io)
            .unwrap();
    };
````

Which feels like the correct way to do this.  I only lend the serial CDC instance into SerialWrapper.

The correct way to implement this would be to add ```embedded_io::ReadReady``` and ```embedded_io::WriteReady```, but those would also need to have access to the ```UsbDevice``` instance to call ```poll(...)``` on, which also feels broken. 

My suggestion would be that using noline in a sync/blocking configuration with an async backend is not a sensible configuration?

The other angle to this argument would be that you would not have a multi-function USB device if you are using noline in a blocking way as this would not work.

Your thoughts??

**Comment by eivindbergem** [comment]
I'm not sure I have much to contribute with here. From what I remember `usbd-serial` didn't really support interrupt-driven usage as it was poll based. What I implemented was mostly a way of working around these limitations. I see that `usbd-serial` has had some released since then – I remember the developer working on a big rewrite – so maybe things have changed in this regard.

In any case, you likely have a better understanding of this at the moment. So, just go ahead and do what you think is the most sensible :)

**Comment by m5p3nc3r** [comment]
Ok, so I have the std[sync/async] and nostd[sync/async] examples working now.

I am moving on to re-integrating the tests.  I have one question for you.  Now that I have moved over to using embedded_io::ErrorKind as the default error type, and we need to enable the std feature in embedded_io to get the From trait implementation for std::io::Error, what is the best way to do this.  I believe there are two options;

1, make std a default feature, this will be a breaking change in the configuration, but it will mean that you can simply run ```carto test``` and have it work as expected.  This also have the benefit of having the intellisense working properly for the test cases.
2, force you to run ```cargo test --features=std``` to enable the embedded_io crate to be build correctly.  The problem with this is that the rust analyser will think there are errors in the test code.

I plan to go with number 1 as I believe it is the most rusty way to do this.  Let me know your thoughts.

**Comment by m5p3nc3r** [comment]
One last question - we have discussed this before, but I wanted to check before I made the change.

Are you Ok with me removing the stm32 example?  I don't have any hardware and am unable to test is.  I have sync and async examples using the rp2040.  Or would somebody with the stm32 board be happy to update the example code?

**Comment by eivindbergem** [comment]
> Ok, so I have the std[sync/async] and nostd[sync/async] examples working now.
> 
> I am moving on to re-integrating the tests. I have one question for you. Now that I have moved over to using embedded_io::ErrorKind as the default error type, and we need to enable the std feature in embedded_io to get the From trait implementation for std::io::Error, what is the best way to do this. I believe there are two options;
> 
> 1, make std a default feature, this will be a breaking change in the configuration, but it will mean that you can simply run `carto test` and have it work as expected. This also have the benefit of having the intellisense working properly for the test cases. 2, force you to run `cargo test --features=std` to enable the embedded_io crate to be build correctly. The problem with this is that the rust analyser will think there are errors in the test code.
> 
> I plan to go with number 1 as I believe it is the most rusty way to do this. Let me know your thoughts.

Another solution here is to use `#![cfg_attr(not(test), no_std)]` to have `std` in tests.

**Comment by eivindbergem** [comment]
> One last question - we have discussed this before, but I wanted to check before I made the change.
> 
> Are you Ok with me removing the stm32 example? I don't have any hardware and am unable to test is. I have sync and async examples using the rp2040. Or would somebody with the stm32 board be happy to update the example code?

You can just remove it for now, and maybe I'll add a new one back in later.

**Comment by m5p3nc3r** [comment]
> > Ok, so I have the std[sync/async] and nostd[sync/async] examples working now.
> > I am moving on to re-integrating the tests. I have one question for you. Now that I have moved over to using embedded_io::ErrorKind as the default error type, and we need to enable the std feature in embedded_io to get the From trait implementation for std::io::Error, what is the best way to do this. I believe there are two options;
> > 1, make std a default feature, this will be a breaking change in the configuration, but it will mean that you can simply run `carto test` and have it work as expected. This also have the benefit of having the intellisense working properly for the test cases. 2, force you to run `cargo test --features=std` to enable the embedded_io crate to be build correctly. The problem with this is that the rust analyser will think there are errors in the test code.
> > I plan to go with number 1 as I believe it is the most rusty way to do this. Let me know your thoughts.
> 
> Another solution here is to use `#![cfg_attr(not(test), no_std)]` to have `std` in tests.

I have noticed that the github actions runner for tests uses ```cargo test --verbose --all-features``` which means that the CI should work.  But if you wanted to run them manually, it would only work if you run the same command - ```cargo test``` would fail.

**Comment by m5p3nc3r** [comment]
@eivindbergem I think it is ready now.  Should I move it ro 'ready for review'?

I appreciate there is a lot of churn in this, but I think its a sensible change.

**Comment by m5p3nc3r** [comment]
@eivindbergem - Thats everything green on the CI now, I don't know how to remove the required test for examples-stm32f103, so it still shows as not completed.  Is this something you can resolve?

**Comment by eivindbergem** [comment]
> @eivindbergem - Thats everything green on the CI now, I don't know how to remove the required test for examples-stm32f103, so it still shows as not completed. Is this something you can resolve?

It says here that one test haven't completed yet, but I don't see it in the list. Not sure what's going on.

**Comment by eivindbergem** [comment]
> > @eivindbergem - Thats everything green on the CI now, I don't know how to remove the required test for examples-stm32f103, so it still shows as not completed. Is this something you can resolve?
> 
> It says here that one test haven't completed yet, but I don't see it in the list. Not sure what's going on.

Figured it out, I had set the stm32 test to required in project settings.

File: examples/no_std/rp2040/.vscode/launch.json
Comment (eivindbergem) [inline examples/no_std/rp2040/.vscode/launch.json:1]: No need for vscode files to be checked in.

**Comment by m5p3nc3r** [comment]
Awesome - let me know if there is anything else you want me to do to the code before merging.

File: noline/src/sync_editor.rs [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+                    InitializerResult::Continue => (),
+                    InitializerResult::Item(terminal) => break terminal,
+                    InitializerResult::InvalidInput => {
+                        return Err(NolineError::ParserError);
+                    }
+                }
+            }
+            if len == 0 {
+                return Err(NolineError::Aborted);
+            }
```
Comment (eivindbergem) [inline noline/src/sync_editor.rs]: This should be if-else.

File: examples/no_std/rp2040-embassy/.vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
+{
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/.vscode/launch.json]: Remove from repo

File: examples/no_std/rp2040-embassy/.vscode/tasks.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/.vscode/tasks.json]: Remove file

File: examples/no_std/rp2040-embassy/README.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Embassy async enabled USB CDC interface
+
+This example code shows how to use the embassy async system to noline in a no_std environment.
+
+It has only been tested on the RP2040, but should be simple to port to other platforms supported by embassy.
+
+It has two async functions running to prove that the two can co-exist.
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/README.md]: I don't think we have anything to prove :stuck_out_tongue: 

Maybe rephrase as "demonstrate".

File: examples/no_std/rp2040-embassy/src/noline_async.rs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        let mut io = MyIO::new(&mut stdin, &mut stdout);
+
+        let mut editor = EditorBuilder::new_static::<64>()
+            .with_static_history::<64>()
+            .build_async(&mut io.0)
+            .await
+            .unwrap();
+
+        while let Ok(line) = editor.readline(prompt, &mut io.0).await {
+            let _ = writeln!(io, "READ: {}", line);
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/src/noline_async.rs]: Won't this block through `FmtWrite::write_str`?

File: examples/no_std/rp2040-embassy/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+panic-probe = { version = "0.3", features = ["print-defmt"] }
+heapless = "0.8.0"
+embedded-hal = { version = "1.0.0" }
+
+# fugit = "0.3.7"
+noline = { path = "../../../noline", default-features = false, features = [
+    "async",
+] }
+nb = "1.0.0"
+fixed-queue = "0.5.1"
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/Cargo.toml:54]: This crate is poorly documented. Is there a better alternative?

File: examples/no_std/rp2040-embassy/src/noline_async.rs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            // Read a maximum of 64 bytes from the ouput
+            let len = self
+                .stdin
+                .read_packet(&mut buf)
+                .await
+                .map_err(|e| map_error(e))?;
+            // This is safe because we only ever pull data when empty
+            // And the queue has the same capacity as the input buffer
+            for i in buf.iter().take(len) {
+                let _ = self.queue.push_back(*i);
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/src/noline_async.rs]: Add `expect()` instead of silently ignoring error.

File: examples/no_std/rp2040-embassy/src/usb.rs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    config.product = Some("USB-serial example");
+    config.serial_number = Some("TEST");
+    config.max_power = 100;
+    config.max_packet_size_0 = 64;
+
+    // Required for windows compatibility.
+    // https://developer.nordicsemi.com/nRF_Connect_SDK/doc/1.9.1/kconfig/CONFIG_CDC_ACM_IAD.html#help
+    config.device_class = 0xEF;
+    config.device_sub_class = 0x02;
+    config.device_protocol = 0x01;
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/src/usb.rs]: Move magic numbers to const's with comments describing their meaning.

File: examples/no_std/rp2040-embassy/src/usb.rs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    config.composite_with_iads = true;
+
+    // Create embassy-usb DeviceBuilder using the driver and config.
+    // It needs some buffers for building the descriptors.
+    //let mut device_descriptor = [0; 256];
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/src/usb.rs]: Remove commented out code

File: examples/no_std/rp2040-embassy/src/usb.rs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    let mut config_descriptor = [0; 256];
+    let mut bos_descriptor = [0; 256];
+    let mut msos_descriptor = [0; 256];
+    let mut control_buf = [0; 64];
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/src/usb.rs]: Use constants for buffer sizes.

File: examples/no_std/rp2040-embassy/src/usb.rs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        config,
+        //&mut device_descriptor,
+        &mut config_descriptor,
+        &mut bos_descriptor,
+        &mut msos_descriptor,
+        &mut control_buf,
+    );
+
+    // Create classes on the builder.
+    let serial = CdcAcmClass::new(&mut builder, &mut state, 64);
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/src/usb.rs]: Use constant for magic buffer size.

File: examples/no_std/rp2040-embassy/src/usb.rs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+    let (mut send, mut recv, mut control) = serial.split_with_control();
+
+    let noline_fut = async {
+        let _ = cli(&mut send, &mut recv, &mut control).await;
```
Comment (eivindbergem) [inline examples/no_std/rp2040-embassy/src/usb.rs]: `cli()` returns `()`, so no need to bind value to `_` as this makes it look like we're ignoring the return value.

File: examples/std/src/bin/std-async-tokio.rs [hunk 13]
```diff
@@ -1,27 +1,51 @@
...
+        let prompt = "> ";
+
+        let mut editor = EditorBuilder::new_unbounded()
+            .with_unbounded_history()
+            .build_async(&mut io.0)
+            .await
+            .unwrap();
+
+        while let Ok(line) = editor.readline(prompt, &mut io.0).await {
+            writeln!(io, "Read: '{}'", line).unwrap();
```
Comment (eivindbergem) [inline examples/std/src/bin/std-async-tokio.rs]: Won't this block?

File: examples/std/src/bin/std-async-tokio.rs [hunk 14]
```diff
@@ -1,27 +1,51 @@
...
         }
-    }
+    });
+
+    let _ = term_task.await;
```
Comment (eivindbergem) [inline examples/std/src/bin/std-async-tokio.rs]: Don't ignore the result.

File: noline/src/async_editor.rs [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+            if len == 1 {
+                match initializer.advance(buf[0]) {
+                    InitializerResult::Continue => (),
+                    InitializerResult::Item(terminal) => break terminal,
+                    InitializerResult::InvalidInput => return Err(NolineError::ParserError),
+                }
+            }
+            if len == 0 {
+                return Err(NolineError::Aborted);
+            }
```
Comment (eivindbergem) [inline noline/src/async_editor.rs]: This should be an if-else since `len` can only be `0` or `1`.

File: noline/src/async_io.rs [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+{
+    /// Create IO wrapper from input and output
+    pub fn new(input: &'a mut R, output: &'a mut W) -> Self {
+        Self { input, output }
+    }
+
+    // /// Consume wrapper and return input and output as tuple
+    // pub fn take(self) -> (R, W) {
+    //     (self.input, self.output)
+    // }
```
Comment (eivindbergem) [inline noline/src/async_io.rs]: Why is this commented out?

File: examples/no_std/rp2040-embassy/src/noline_async.rs (hunk 5 above)
Comment (m5p3nc3r) [reply examples/no_std/rp2040-embassy/src/noline_async.rs]: I don't believe so because the async executor in embassy will take care of the 'would block' condition on the USB serial interface.

File: examples/std/src/bin/std-async-tokio.rs (hunk 13 above)
Comment (m5p3nc3r) [reply examples/std/src/bin/std-async-tokio.rs]: I dont think so because I changed the executor to be multi threaded, and this is running in its own context.
I believe this should be Ok.

File: examples/no_std/rp2040-embassy/Cargo.toml (hunk 6 above)
Comment (m5p3nc3r) [reply examples/no_std/rp2040-embassy/Cargo.toml:54]: I could try using this instead https://crates.io/crates/mini-io-queue. Lrt me know if this works for you.

Comment (m5p3nc3r) [reply examples/no_std/rp2040-embassy/Cargo.toml:54]: Had a look, but I think the application of the crate is very simple, and its only in an example, not the core of noline.  Can I suggest we keep fixed-queue for now?

File: examples/std/src/bin/std-async-tokio.rs (hunk 13 above)
Comment (eivindbergem) [reply examples/std/src/bin/std-async-tokio.rs]: Even if the executor is multi-threaded we shouldn't block in async code. The original code without `writeln!` was fine, so I don't really see the reason for changing it.

File: examples/no_std/rp2040-embassy/src/noline_async.rs (hunk 5 above)
Comment (eivindbergem) [reply examples/no_std/rp2040-embassy/src/noline_async.rs]: Maybe I'm missing something here. How does embassy take care of this exactly? `writeln!()` results in a call to `write_str()` which in turn calls `block_on()` which will block the calling thread, which in this context is the async task which will block the entire embassy executor.

File: examples/no_std/rp2040-embassy/Cargo.toml (hunk 6 above)
Comment (eivindbergem) [reply examples/no_std/rp2040-embassy/Cargo.toml:54]: If it's too much trouble, we can keep it.

File: examples/std/src/bin/std-async-tokio.rs (hunk 13 above)
Comment (m5p3nc3r) [reply examples/std/src/bin/std-async-tokio.rs]: Happy to revert it, I was trying to ensure that the application end of the code - the bit that was using noline, was as similar as possible across all 4 examples.

Comment (m5p3nc3r) [reply examples/std/src/bin/std-async-tokio.rs]: Done.

File: examples/no_std/rp2040-embassy/src/noline_async.rs (hunk 5 above)
Comment (m5p3nc3r) [reply examples/no_std/rp2040-embassy/src/noline_async.rs]: Ok, turns our my understanding of the async executor is broken under block_on usage.  See https://docs.rs/embassy-futures/latest/embassy_futures/fn.block_on.html
My bad - I'll fix this in the same way as the Tokio example.  Thanks for spotting this.

Comment (m5p3nc3r) [reply examples/no_std/rp2040-embassy/src/noline_async.rs]: Fixed now.

**Comment by eivindbergem** [comment]
It looks very good now! I can squash and merge now, if you're ready.

**Comment by m5p3nc3r** [comment]
Sure - I'm already using it in my embassy project, all seems fine :o)
